Add some bench targets

fyrchik opened the pull request:
Close #5, #78,

TODO: code cleanup

roman-khimov commented on cmd/internal/gen.go:
```diff
@@ -56,42 +63,56 @@ var SourceWifs = []string{
-		contractHash, "transfer", callflag.All,
-		fromAddressHash, fromAddressHash, int64(1), nil)
-	emit.Opcodes(w.BinWriter, opcode.ASSERT)
+	for i := 0; i < n; i++ {
```
That's certainly not the way it was supposed to work. Single transaction should still have a single transfer inside, it's just that instead of using one address in every transaction we use a number of them.

roman-khimov commented on cmd/internal/gen.go:
```diff
@@ -47,10 +47,19 @@ const (
 )
 
+// SourceWifs represents the senders of NEO and GAS.
+var SourceWifs = []string{
```
How about 1000 or 10K addresses?

fyrchik replied:
Currently not supported.
The problem is that we transfer funds in chain dump, but send during tx generation. WIFs must be the same.
So we either should store all 1000 wifs in code/file or generate them on the fly in deterministic fashion.

roman-khimov replied:
But that's what we actually need, storing them in a file is not a big deal. And if we're to solve #84 it'd be easy to do initial transfers using dump's data.

roman-khimov commented on README.md:
```diff
@@ -324,6 +324,7 @@ Name|Description|Default|Example
 NEOBENCH_LOGGER|Container logging facility|`none`|`none`, `journald`, `syslog`
 NEOBENCH_TC|Parameters passed to the `tc qdisc` (netem discipline) on container startup| |`delay 100ms`
 NEOBENCH_TYPE|Type of the load|`NEO`|`NEO`, `GAS`, `NEO-2-3` (transfer NEO from 2 senders to 3 recepients)
+NEOBENCH_VOTE|Vote for validators before the bench|empty|`1` or empty
```
I'm not sure what's the difference. Registration only vs. registration with votes? Does it change anything? Can we always do it?

fyrchik replied:
Registration is done for `VOTE` load to be applicable. `Vote` in dump is done for the `Transfer` loads.

roman-khimov replied:
I mean you can always register some candidates, it doesn't hurt.

AnnaShaleva commented on cmd/dump/main.go:
```diff
@@ -234,6 +234,40 @@ func fillChain(bc *core.Blockchain, c *signer) error {
+		callflag.All, p.PublicKey().Bytes())
+	if w.Err != nil {
+		panic(w.Err)
+	}
```
Maybe it's worth to add an `ASSERT`?

fyrchik commented:
I have rebased this on master:
1. Added some info to the tx dump header, wifs are generated on the fly.
2. Remove bench for `vote` call performance.
3. More appropriate parameter handling (3 parameters instead of `NEO-10-22`).

TODO: test with sharp RPC.

fyrchik commented:
Checked C# RPC with voting and custom sender count, everything looks good.

roman-khimov commented on cmd/gen/main.go:
```diff
@@ -13,6 +13,9 @@ var (
 	typ = flag.String("type", internal.NEOTransfer, "Type of txs that would be generated.")
+
+	fromCount = flag.Int("from", 1, "Amount of tx senders")
+	toCount   = flag.Int("to", 1, "Amount of tx recepients")
```
This actually changes 1-1 from using the same account in `from` and `to` to using two separate accounts.

fyrchik replied:
Fixed. Now first receivers are taken from senders and others are deterministically generated.

roman-khimov commented on cmd/internal/settings.go:
```diff
@@ -94,6 +94,8 @@ func InitSettings() *viper.Viper {
 		"``Path to input file to load transactions.\n"+
 			"Example: -i ./dump.txs --in /path/to/import/transactions")
 
+	flags.BoolP("vote", "", false, "Vote befor the bench.")
```
before

fyrchik commented on cmd/gen/main.go:
```diff
@@ -23,7 +24,12 @@ func main() {
 		internal.ReadDump(*inp)
 	case out != nil && *out != "" && cnt != nil && *cnt > 0:
-		internal.WriteDump(ctx, *out, *typ, *cnt)
+		p, _ := keys.NewPrivateKeyFromWIF("KxhEDBQyyEFymvfJD96q8stMbJMbZUb6D1PmXqBWZDU2WvbvVs9o")
```
Yes, this key is used in the deployed contract.

fyrchik commented on cmd/internal/gen.go:
```diff
@@ -129,7 +139,8 @@ func Generate(ctx context.Context, opts BenchOptions, callback ...GenerateCallba
 			}
 
-			txCh <- tx
+			index := rand.Intn(len(txR))
```
Made this more deterministic.
I don't it collecting them in order is important, as we have some randomness during tx send already (we send transactions in parallel).

roman-khimov replied:
Still, initial order can affect the result (as transactions are read and tried to be sent in order) and it'll probably be a bit different from run to run (and on different machines). Some bias could be added by these variations, so we better avoid it.

roman-khimov reviewed (commented):
Tx generation non-determinism still worries me.

fyrchik commented on cmd/internal/gen.go:
```diff
@@ -96,30 +101,52 @@ func Generate(ctx context.Context, opts BenchOptions, callback ...GenerateCallba
+		max = opts.ToCount
+	}
+
+	txR := make([]txRequest, max)
```
Sending funds from multiple accounts? I don't think some symmetry is essential here (and if needed it can be achieved with different parameters, e.g. 6/3 or 6/12).

fyrchik commented:
Implemented deterministic generation in the last commit (checked with `sha256sum` for `GAS.1.1` dump).
